packages/core/src/operator/listenForChallenges.ts

```diff
@@ -11,40 +9,55 @@ import { resilientEventListener } from "../utils/resilientEventListener.js";
  * @returns A function that can be called to stop listening for the event.
  */
-export function listenForChallenges(callback: (challengeNumber: bigint, challenge: Challenge, event: any) => void): () => void {
-
-    // create a map to keep track of challengeNumbers that have called the callback
+export function listenForChallenges(callback: (challengeNumber: bigint, challenge: Challenge, event: any) => void, onError?: (error: Error) => void): () => void {
     const challengeNumberMap: { [challengeNumber: string]: boolean } = {};
 
-    // listen for the ChallengeSubmitted event
     const listener = resilientEventListener({
         rpcUrl: config.arbitrumOneWebSocketUrl,
         contractAddress: config.refereeAddress,
         abi: RefereeAbi,
         eventName: "ChallengeSubmitted",
         log: console.info,
         callback: async (log, error) => {
-
-            if(error){
-                //TODO hanlde on error
+            if (error) {
+                if(!onError) return;
+                // Call the onError function with the caught error
+                onError(error instanceof Error ? error : new Error(String(error)));
                 return;
             }
+            try {
+                const challengeNumber = BigInt(log?.args[0]);
 
-            const challengeNumber = BigInt(log?.args[0]);
-
-            // if the challengeNumber has not been seen before, call the callback and add it to the map
-            if (!challengeNumberMap[challengeNumber.toString()]) {
-                challengeNumberMap[challengeNumber.toString()] = true;
+                // if the challengeNumber has not been seen before, call the callback and add it to the map
+                if (!challengeNumberMap[challengeNumber.toString()]) {
+                    challengeNumberMap[challengeNumber.toString()] = true;
 
-                // lookup the challenge
-                const challenge = await getChallenge(challengeNumber);
+                    // lookup the challenge
+                    const challenge = await getChallenge(challengeNumber);
 
-                void callback(challengeNumber, challenge, log);
+                    void callback(challengeNumber, challenge, log);
+                }
+            } catch (err) {
+                if (!onError) return;
+            
+                const errorContext = {
+                    message: err instanceof Error ? err.message : String(err),
+                    stack: err instanceof Error ? err.stack : "No stack available",
+                    context: {
+                        eventName: "ChallengeSubmitted",
+                        functionName: "listenForChallenges",
+                    },
+                    timestamp: new Date().toISOString()
+                };
+            
+                // Log the error with additional context
+                console.error("An error occurred:", errorContext);
```

Comment:
Please remove that `errorContext`, I have no idea what this is doing and it seems like redundant mapping of the error object.

```
} catch (err: any) 

console.error("There was an error trying to read the current challenge (${challengeNumber}) - Error: ${error && error.message ? error.message : error}`)`

```
Then make it so the log only gets printed when the `onError` is undefined, if the onError is passed in we should expect the onError to handle the console log.

Reply:
@XDapps Please update that and then ping me so we can get this merged

Reply:
@CryptITAustria this has been updated per your instructions! Thank you